lib/protocol/protocol.go

```diff
@@ -813,44 +819,50 @@ func (c *rawConnection) shouldCompressMessage(msg message) bool {
 	}
 }
 
-// Close is called when the connection is regularely closed and thus the Close
-// BEP message is sent before terminating the actual connection. The error
-// argument specifies the reason for closing the connection.
+// Close is called when the connection is closed while (presumably) still working
+// and thus the Close BEP message is sent before terminating the actual
+// connection. The error argument specifies the reason for closing the connection.
 func (c *rawConnection) Close(err error) {
-	c.sendCloseOnce.Do(func() {
+	c.closeOnce.Do(func() {
+		c.stopAndCleanup(err)
+
 		done := make(chan struct{})
-		c.send(&Close{err.Error()}, done)
-		select {
-		case <-done:
-		case <-c.closed:
-		}
-	})
+		c.writeMessage(asyncMessage{&Close{err.Error()}, done})
+		<-done
 
-	// No more sends are necessary, therefore further steps to close the
-	// connection outside of this package can proceed immediately.
-	// And this prevents a potential deadlock due to calling c.receiver.Closed
-	go c.internalClose(err)
+		// No more sends are necessary, therefore further steps to close the
+		// connection outside of this package can proceed immediately.
+		// And this prevents a potential deadlock due to calling c.receiver.Closed
+		go c.receiver.Closed(c, err)
+	})
 }
 
-// internalClose is called if there is an unexpected error during normal operation.
+// internalClose is called if there is an error and sending doesn't work anymore
 func (c *rawConnection) internalClose(err error) {
 	c.closeOnce.Do(func() {
-		l.Debugln("close due to", err)
-		close(c.closed)
-
-		c.awaitingMut.Lock()
-		for i, ch := range c.awaiting {
-			if ch != nil {
-				close(ch)
-				delete(c.awaiting, i)
-			}
-		}
-		c.awaitingMut.Unlock()
+		c.stopAndCleanup(err)
```

Comment:
This will deadlock, because internalClose is called from the writer loop and stopAndCleanup will wait for the writer loop to exit... I'm not sure that's the only deadlock possibility, I didn't look closely, but I'm suspicious in general :)

Reply:
Right, I assumed (...) that once `sync.Once.Do` was invoked, any subsequent calls will just skip it, but documentation says your suspicion is right, it blocks all subsequent calls until it returns. I didn't see an obvious way to fix this with `Once`, so I did it with known means of a channel and a lock.